pkg/controller/replication_controller.go

```diff
@@ -115,181 +91,276 @@ func NewReplicationManager(kubeClient client.Interface) *ReplicationManager {
+	for {
+		func() {
+			key, quit := rm.queue.Get()
+			if quit {
+				return
+			}
+			defer rm.queue.Done(key)
+			err := rm.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing replication controller: %v", err)
+			}
+		}()
 	}
-	return false
 }
 
-func (rm *ReplicationManager) syncReplicationController(controller api.ReplicationController) error {
-	s := labels.Set(controller.Spec.Selector).AsSelector()
-	podList, err := rm.kubeClient.Pods(controller.Namespace).List(s, fields.Everything())
-	if err != nil {
-		return err
-	}
-	filteredList := filterActivePods(podList.Items)
-	numActivePods := len(filteredList)
-	diff := numActivePods - controller.Spec.Replicas
+// manageReplicas for the given replication controller.
+func (rm *ReplicationManager) manageReplicas(filteredPods []*api.Pod, controller api.ReplicationController) {
+	diff := len(filteredPods) - controller.Spec.Replicas
 	if diff < 0 {
 		diff *= -1
+		rm.expectations.SetExpectations(&controller, diff, 0)
 		wait := sync.WaitGroup{}
 		wait.Add(diff)
 		glog.V(2).Infof("Too few %q replicas, creating %d", controller.Name, diff)
 		for i := 0; i < diff; i++ {
 			go func() {
 				defer wait.Done()
-				rm.podControl.createReplica(controller.Namespace, controller)
+				if err := rm.podControl.createReplica(controller.Namespace, controller); err != nil {
+					rm.expectations.LowerExpectations(controller, 1, 0)
+					util.HandleError(err)
+				}
 			}()
 		}
 		wait.Wait()
 	} else if diff > 0 {
+		rm.expectations.SetExpectations(&controller, 0, diff)
 		glog.V(2).Infof("Too many %q replicas, deleting %d", controller.Name, diff)
 		// Sort the pods in the order such that not-ready < ready, unscheduled
 		// < scheduled, and pending < running. This ensures that we delete pods
 		// in the earlier stages whenever possible.
-		sort.Sort(activePods(filteredList))
+		sort.Sort(activePods(filteredPods))
 
 		wait := sync.WaitGroup{}
 		wait.Add(diff)
 		for i := 0; i < diff; i++ {
 			go func(ix int) {
 				defer wait.Done()
-				rm.podControl.deletePod(controller.Namespace, filteredList[ix].Name)
+				if err := rm.podControl.deletePod(controller.Namespace, filteredPods[ix].Name); err != nil {
+					rm.expectations.LowerExpectations(controller, 0, 1)
+				}
 			}(i)
 		}
 		wait.Wait()
 	}
-	if controller.Status.Replicas != numActivePods {
-		controller.Status.Replicas = numActivePods
-		_, err = rm.kubeClient.ReplicationControllers(controller.Namespace).Update(&controller)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
 }
 
-func (rm *ReplicationManager) synchronize() {
-	// TODO: remove this method completely and rely on the watch.
-	// Add resource version tracking to watch to make this work.
-	var controllers []api.ReplicationController
-	list, err := rm.kubeClient.ReplicationControllers(api.NamespaceAll).List(labels.Everything())
+// syncReplicationController if it's fulfilled its expectations. Update it's Status even if it hasnt't.
```

Comment:
nit: I find this style of using the function name as part of the sentence to be clever but also pretty jarring to read-- this may be clearer if you write out in other words what the function does.